RFD 83: Add license warning RFD (#15766)
gravitational/teleport: rfd/0083-license-expiration-warnings.md

```

CLI does not support snoozing the warnings as you can in the webUI. Which provides the functionality to disable the warnings temporarily.

### Web UI UX

Web UI will display license expiration banner on top of the page with the following rules:
Show warning (yellow) to users N days prior to expiration, 1 day snooze, based on license_warning_days role option with same warning text as the CLI
Show error (red) to all users after expiration, no dismiss or snoozing available.
Snoozing allows users to disable the warning in the web UI from being displayed for N days.

### Configuration UX

The time interval in which to show a warning before expiry can be configured on a per role basis.
Interval can be configured by modifying a role option ‘license_warning_days’.
Default will be 90 days, admins can configure if needed.
Default preset roles will use the default. For example the editor/auditor/access roles.
When multiple roles for the same user have different values for this interval the longest one will be used.
This value cannot be configured to be below the minimum value of 10 days.

```yaml
kind: role
version: v5
metadata:
  name: editor
spec:
  options:

>     license_warning_days: 90  # ← defaults to 30, min 10

Reviewer: @EdwardDowling Let's remove configurable time interval from scope as discussed with @xinding33.

Author: Sounds good to me, this should simplify everything.

```

Default values for default roles:
Editor - 90 days
Auditor - 30 days
Access - 30 days

If you have multiple roles, the longest time period takes precedence.

## Implementation details

Existing web api endpoint ‘/ping’ already returns a license warning. Due to this endpoint being unauthenticated and the proposed functionality requires the ability to differentiate based on identity/role this cannot be used as is.
Instead this RFD proposes a new authenticated endpoint ‘/license’ be exposed to provide the required functionality. This would consist of
- An auth api method under the hood that checks the license expiration and determines the warning to give.
- A webapi endpoint that calls it.
  - Method: GET

>   - Path: `/webapi/check-license`

Reviewer: We probably don't need a separate endpoint now that we've decided to not do configurable intervals, and can just piggyback on the cluster alerts endpoint.

@hatched You were working on exposing cluster alerts API IIRC - is there a PR or a branch?

Reviewer 2: The UI changes are in [this branch](https://github.com/gravitational/webapps/tree/hatched/banner) and I'm waiting on [the upgrade alerts pr](https://github.com/gravitational/teleport/pull/15774) to land and then I'll create a new pr with the routes and logic for the apiserver. I have it all working locally and can update this with PR's once ready.

Here is an example of the data structure that's being returned:

```json
{
  "alerts": [
    {
      "kind": "cluster_alert",
      "version": "v1",
      "metadata": {
        "name": "upgrade-suggestion",
        "labels": {
          "teleport.internal/alert-on-login": "yes",
          "teleport.internal/alert-permit-all": "yes"
        },
        "expires": "2022-08-31T17:26:05.728149Z"
      },
      "spec": {
        "severity": 5,
        "message": "A new major version of Teleport is available. Please consider upgrading your cluster to v10.",
        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

```

```
$ build/tsh login --proxy=localhost:9876 --user someUser
> Profile URL:        https://localhost:9876
  Logged in as:       someUser
  Cluster:            localhost
  Roles:              access, editor
  Logins:             someUser
  Valid until:        2022-08-16 03:38:07 +0100 IST [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

  Your Teleport Enterprise Edition license has expired. Please reach out to your
  Teleport Account Manager (or licenses@goteleport.com) to obtain a new license.
```

```
$ build/tctl status -c /home/edward/teleport/teleport.yaml
Cluster  localhost                                                               
Version  10.1.2                                                                  
host CA  never updated                                                           
user CA  never updated                                                           
db CA    never updated                                                           
jwt CA   never updated                                                           
CA pin   sha256:3d72102f020146d09ff400810f59f70b8163ebfe6ec1ecfb0b3b2a0c151592

> Your Teleport Enterprise Edition license will expire in 10 days. Please reach out
> to   your Teleport Account Manager (or licenses@goteleport.com) to obtain a new
> license.

Reviewer 3: ```suggestion
Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support."
```

---

> authors: edwarddowling (edward.dolwing@goteleport.com), r0mant (roman@goteleport.com)

Reviewer: ```suggestion
authors: edwarddowling (edward.dowling@goteleport.com), r0mant (roman@goteleport.com)
```

---

# RFD 83 - License warnings

## Required approvers

- Engineering: `@r0mant`
- Product: `@klizentas && @xinding33`
- Security: `@reedloden`

## What

This RFD proposes a way to display license warnings to self hosted users both before and after license expiration.

## Why

To encourage the renewal of licenses after expiration and let users know before their license expires.

## Success criteria

Teleport displays license warnings to users in both CLI and web UI starting from a certain time prior to the license expiration, as well as post-expiration.

> Teleport administrators can optionally configure the license expiration warning interval but not disable it completely on a per role basis.

Reviewer: @xinding33 To piggyback on this - is having a configurable license warning interval per-role a hard requirement from your standpoint? It's fine if it is - just wanted to double-check since having the same warning interval for everyone would definitely simplify implementation and it's another feature to support.

Also, presumably cluster admins can just set it to the minimum of 10 days for everyone to "suppress" the default behavior which would sort of alleviate our attempts to make it more/less annoying for different types of users.

Reviewer 3: @r0mant Configurable is nice to have but not a hard requirement. As discussed, it's hard to determine who "admins" are and who "regular users" are so that's why we suggested a role option instead of tying it arbitrarily to existing role options or roles.

Are you suggesting we remove the role option entirely and just alert everyone N days ahead? Or are you suggesting we change the role option from integer to boolean?

Reviewer: @xinding33 The former:

> remove the role option entirely and just alert everyone N days ahead

If that works for you, that would simplify the implementation significantly and allow us to reuse the same cluster alerts mechanism we'll be using in automatic upgrade notifications.

## Scope

Self-hosted Teleport will be the target of this RFD. Teleport Cloud is out of scope because the Cloud license does not have proper expiration yet. When Cloud implements proper license expiration, this feature will kick in automatically.
License warning should be displayed in both Web UI and CLI (tsh/tctl).

### CLI UX

License warnings should be displayed in CLI (tsh/tctl) 
TSH - on “tsh login”, in “tsh status”
When the user uses either the “tsh login” or ‘tsh status” commands the appropriate license warning will be displayed. (see examples below)
TCTL - “tctl status”
When the user uses the “tctl status” command the appropriate license warning will be displayed.

> After expiration all tsh and tctl will display the license expiry warning.

Reviewer: > all tsh and tctl will display the license expiry warning

All commands?

### Example outputs

The license warning portion of the outputs should be coloured red and orange respectively for the expired and unexpired warnings.

```bash
$ build/tsh status
> Profile URL:        https://localhost:9876
  Logged in as:       someUser
  Cluster:            someCluster
  Roles:              access, editor
  Logins:             someUser
  Valid until:        2022-08-12 21:57:19 +0100 IST [valid for 4h44m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
  Your Teleport Enterprise Edition license will expire in 10 days. Please reach out
  to   your Teleport Account Manager (or licenses@goteleport.com) to obtain a new

>   license.

Reviewer: @xinding33 Can you provide a copy for the warning message we want to use here?

Reviewer 3: "Your Teleport Enterprise Edition license will expire in N days. Please reach out to licenses@goteleport.com to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support."

  Your Teleport Enterprise Edition license has expired. Please reach out to your
  Teleport Account Manager (or licenses@goteleport.com) to obtain a new license.
```

```
$ build/tctl status -c /home/edward/teleport/teleport.yaml
Cluster  localhost                                                               
Version  10.1.2                                                                  
host CA  never updated                                                           
user CA  never updated                                                           
db CA    never updated                                                           
jwt CA   never updated                                                           
CA pin   sha256:3d72102f020146d09ff400810f59f70b8163ebfe6ec1ecfb0b3b2a0c151592

Your Teleport Enterprise Edition license will expire in 10 days. Please reach out
to   your Teleport Account Manager (or licenses@goteleport.com) to obtain a new
license.
```

CLI does not support snoozing the warnings as you can in the webUI. Which provides the functionality to disable the warnings temporarily.

### Web UI UX

> For web UI - banner on top

Reviewer: ```suggestion
Web UI will display license expiration banner on top of the page with the following rules:
```

Reviewer: Also, can you link a ticket with the design we had here?

Show warning (yellow) to users N days prior to expiration, 1 day snooze, based on license_warning_days role option with same warning text as the CLI
Show error (red) to all users after expiration, no dismiss or snoozing available.
Snoozing allows users to disable the warning in the web UI from being displayed for N days.

### Configuration UX

The time interval in which to show a warning before expiry can be configured on a per role basis.
Interval can be configured by modifying a role option ‘license_warning_days’.
Default will be 30 days, admins can configure if needed.

> Default preset roles will use the default. For example the editor/auditor/access roles.

Reviewer: I think we agreed on setting it to 90 days for the editor role.

> When multiple roles for the same user have different values for this interval the shortest one will be used.

Reviewer: Wouldn't we want to use the longest rather?

This value cannot be configured to be below the minimum value of 10 days.

> Idea: Keep this setting as a role option

Reviewer: ```suggestion
```

```yaml
kind: role
version: v5
metadata:
  name: editor
spec:
  options:
    license_warning_days: 90  # ← defaults to 30, min 10
```

Default values for default roles:
Editor - 90 days
Auditor - 30 days
Access - 30 days

> If you have multiple roles, the shortest time period takes precedence.

Reviewer: ```suggestion
If you have multiple roles, the longest time period takes precedence.
```

## Implementation details

Existing web api endpoint ‘/ping’ already returns a license warning. Due to this endpoint being unauthenticated and the proposed functionality requires the ability to differentiate based on identity/role this cannot be used as is.

> Instead this RFD proposes a new authenticated endpoint ‘/license’ be exposed to provide the required functionality. This would consist of

Reviewer: Can you document the full reference for the new endpoint? The full path, request, response in a more structured way. Something like (but clean it up):

```
GET /webapi/...../check-license

Request: empty
Response: {"messages": ["..."]}
```

Reviewer 4: FYI, @fspmarshall's work introduces a new backend resource called a "cluster alert" that is to be used to notify the user when an upgrade is available. @hatched is working on a new endpoint to expose these alerts.

I think a license expired or about to expire also qualifies as a cluster alert.

Can we look to combine these two approaches so we don't have two separate ways to present alerts to the user?

cc @hatched @r0mant @xinding33

Reviewer: I'm not sure we'll be able to reuse the same mechanism because the license warnings will be different based on the identity of the user, users can have different intervals of when the warning will be shown to them. We'll see what we can reuse, if anything.

Reviewer: Maybe we can reuse the same data structures and API at least, so if there will be an API for retrieving cluster alerts, the handler can also look at whether there are license warnings for a particular user who's making this request and enrich the "cluster alert" with the additional license warning before returning it to the user. This I think would make sense.

Reviewer: @hatched Is there a PR for exposing these alerts to users which we can piggyback on? @EdwardDowling Let's update the RFD with this.

Reviewer 2: The UI is actually using `GET /enterprise/license/status` not `GET /ping` to get this data. This endpoint is also unauthenticated.

Reviewer: If it's unauthenticated, then we won't be able to use it to implement the requirements in this RFD.

Reviewer 5: > I'm not sure we'll be able to reuse the same mechanism because the license warnings will be different based on the identity of the user, users can have different intervals of when the warning will be shown to them.

FWIW the "cluster alerts" decide if a user is allowed to see a given alert based on labels.  If the logic around which users should see which license warnings is fairly straightforward, you might just be able to add a new label and update `ServerWithRoles.GetClusterAlerts` to do some special checks if it sees that label.

An auth api method under the hood.
A webapi endpoint that calls it.
Its request body would be empty and the the response would be of the form

```
message LicenseWarning { 

>     WarningMessage String

Reviewer: Should probably be a list of strings so we can return multiple messages.

}
```

## Security considerations

> Admins cannot fully disable warnings, because there will be a minimum threshold of 30 days.

Reviewer: I think RFD mentions that the minimum would be 10 days above.

```

```
$ build/tctl status -c /home/edward/teleport/teleport.yaml
Cluster  localhost                                                               
Version  10.1.2                                                                  
host CA  never updated                                                           
user CA  never updated                                                           
db CA    never updated                                                           
jwt CA   never updated                                                           
CA pin   sha256:3d72102f020146d09ff400810f59f70b8163ebfe6ec1ecfb0b3b2a0c151592

Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.
```

CLI does not support snoozing the warnings as you can in the webUI. Which provides the functionality to disable the warnings temporarily.

### Web UI UX

Web UI will display license expiration banner on top of the page with the following rules:
Show warning (yellow) to users N days prior to expiration, 1 day snooze

> Show error (red) to all users after expiration, no dismiss or snoozing available.

Reviewer 2: Note for how the UI interprets severity integer values:
| Severity | Value | Color |
|---|---|---|
| low | 0 - 4 | blue |
| medium | 5 - 9 | yellow |
| high | 10 + | red |

      }
    }
  ]
}
```

```
{
  "alerts": [
    {
      "kind": "cluster_alert",
      "version": "v1",
      "metadata": {
        "name": "123e4567-e89b-12d3-a456-426614174000",
        "labels": {
          "teleport.internal/alert-on-login": "yes",
          "teleport.internal/alert-permit-all": "yes",
          "teleport.internal/license-warning-expired": "yes",
        },
        "expires": "2022-08-31T17:26:05.728149Z"
      },
      "spec": {
        "severity": 5,
        "message": "Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.",
        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

On startup and every 1 hour afterwards the auth server will check the license and generate a license warning if applicable, or clear license warning alerts if they are no longer needed.

These license warning alerts will need to be auth server specific so the host id will need to be added to the cluster alert spec to allow this.

## Security considerations

> If your license expired, you can’t disable or snooze the warning.

Reviewer 2: We will need to modify the `alert` spec to include a `dismissible` boolean flag. 

We can cheat it for now to say something like `high` severity alerts are not dismissible but I think adding a flag would be more resilient.

Author: Could the non dismissible alerts be determined by a label?

Author: Discussed this over slack with hatched, going to go with the boolean flag

        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

```
{
  "alerts": [
    {
      "kind": "cluster_alert",
      "version": "v1",
      "metadata": {
        "name": "123e4567-e89b-12d3-a456-426614174000",
        "labels": {
          "teleport.internal/alert-on-login": "yes",
          "teleport.internal/alert-permit-all": "yes",
          "teleport.internal/license-warning-expired": "yes",
        },
        "expires": "2022-08-31T17:26:05.728149Z"
      },
      "spec": {
        "severity": 5,
        "message": "Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.",
        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

On startup and every 1 hour afterwards the auth server will check the license and generate a license warning if applicable, or clear license warning alerts if they are no longer needed.

> These license warning alerts will need to be auth server specific so the host id will need to be added to the cluster alert spec to allow this. The cluster alert spec may also need to be modified to add a bool for wether an alert is allowed to be dismissed.

Reviewer 2: ```suggestion
These license warning alerts will need to be auth server specific so the host id will need to be added to the cluster alert spec to allow this. The cluster alert spec may also need to be modified to add a bool for whether an alert is allowed to be dismissed.
```

> These license warning alerts will need to be auth server specific so the host id will need to be added to the cluster alert spec to allow this. The cluster alert spec may also need to be modified to add a bool for wether an alert is allowed to be dismissed.

Reviewer: ```suggestion
These license warning alerts will need to be auth server specific so the host id will need to be added to the cluster alert spec to allow this. The cluster alert spec may also need to be modified to add a bool for whether an alert is allowed to be dismissed.
```

---
authors: edwarddowling (edward.dowling@goteleport.com), r0mant (roman@goteleport.com)
---

# RFD 83 - License warnings

## Required approvers

- Engineering: `@r0mant`
- Product: `@klizentas && @xinding33`
- Security: `@reedloden`

## What

This RFD proposes a way to display license warnings to self hosted users both before and after license expiration.

## Why

To encourage the renewal of licenses after expiration and let users know before their license expires.

## Success criteria

Teleport displays license warnings to users in both CLI and web UI starting from a certain time prior to the license expiration, as well as post-expiration.

## Scope

Self-hosted Teleport will be the target of this RFD. Teleport Cloud is out of scope because the Cloud license does not have proper expiration yet. When Cloud implements proper license expiration, this feature will kick in automatically.
License warning should be displayed in both Web UI and CLI (tsh/tctl).

### CLI UX

License warnings should be displayed in CLI (tsh/tctl) 

> TSH - on “tsh login”, in “tsh status”

Reviewer 5: `tsh status` currently makes no network calls, using only the local filesystem to generate its output.  This can be very nice for folks with slower networks, since it gives them a fast way to get key information at a glance.

When Implementing version alerts I opted not to show them in `tsh status` for this reason.  IMO if license warnings are to be displayed in `tsh status`, they should be behind a flag (e.g. `-v/--verbose`) in order to preserve the snappiness of the current `tsh status` impl by default.

Author: Not including these alerts in the standard `tsh status` makes sense to me. There is a comment in tsh.go mentioning the possibility of adding a `tsh alerts ls` command eventually which we could use instead of the `tsh status -v` as an option for viewing alerts.

Although when the license has expired we would like to show the warnings on every command. We could exclude the `tsh status` command from this to keep it snappy. What are your thoughts on this @xinding33

Reviewer: Let's see what @xinding33 says but I'm pretty sure we do want to show warnings on `tsh status` by default, esp. if your license has expired. We can cache the alerts on the client somehow?

Reviewer: Thinking more about this, maybe caching on the client is not a great idea since we'll run into problem with invalidation if someone has updated the license while you're logged in, so might get unnecessarily complicated.

@fspmarshall I'm not sure tbh we'll be able to maintain this properly that `tsh status` doesn't make network calls to the cluster. There are alerts that we definitely do want to show on status (e.g. these license ones). Hiding them behind `-v` flag is I think not much better than not showing them at all as I'm sure people don't run verbose status very often and we do want these alerts to be seen.

@xinding33 It's a product call at the end of the day so let us know what you think.

Reviewer 3: My perspective is that we need this check in `tsh status` (and not put it behind verbose flag) when tsh is connecting to an enterprise cluster to support our pricing model. Latency is either acceptable or unacceptable. In the case that it's acceptable, adding it to `tsh status` won't be detrimental to use experience and in the case that it's unacceptable, we need to fix it because even if we don't add it to `tsh status`, high latency will impact other use cases. What's the expected added latency to `tsh status` based on the current suggested implementation? Typically, anything less than 100ms is not noticeable.

Reviewer 2: We could add a timeout to the request so that in the event the request takes longer than say 350ms it times out and dumps the status as normal. In this case the status output could have a little note that says: 

```
license check timed out, run: the following to check your license information:
tsh check-license
```

This way we get the best of both worlds.

Author: @fspmarshall  @xinding33 are both of you satisfied with the timeout approach? If so i can update the design doc and add it into the implementation i'm working on.

Reviewer 5: Timeout approach seems reasonable to me.  How about instead of querying specifically for license alerts, we just query for all *high severity* alerts.  If we're going to show alerts on `tsh status` we might as well show security patch alerts too.

Reviewer: This makes sense to me. @EdwardDowling Let's update the RFD to address these two things:

- High severity alerts will be shown on tsh status.
- tsh status will use some reasonable timeout (say, 300-500ms) for fetching alerts.

And merge.

Reviewer: @rosstimothy @fspmarshall I hear you guys and in general I agree with the sentiment. On the other hand, I'm not really sure what specific problem adding an extra call would introduce here, besides increased latency which in the worst-case scenario will make the status command feel not instant but still snappy. Where does the requirement of not having `tsh status` make calls to the cluster come from? I think it's just a happenstance that it doesn't make them now and wasn't necessarily a conscious design decision (unless I'm mistaken).

We could in theory fetch all alerts on login and cache them locally, but then we'd run into problems with invalidation, etc which sounds like over-engineering for something as simple as showing alerts so I'd like to avoid this approach.

If we want to be able to show key information without waiting on network, we can also consider displaying current "local" status right away, while we're fetching alerts in the background. Then the command will exit after it's received/displayed alerts (or upon timeout). Not sure if it's better or worse UX though.

WDYT?

Author: So is the current plan to only show the alert on login and after every tctl once the license expires? Leaving the tsh commands untouched for?

Author: @r0mant @rosstimothy  could confirm the most recent changes to the RFD are what you were thinking I can go ahead and merge this?

- Security: `@reedloden`

## What

This RFD proposes a way to display license warnings to self hosted users both before and after license expiration.

## Why

To encourage the renewal of licenses after expiration and let users know before their license expires.

## Success criteria

Teleport displays license warnings to users in both CLI and web UI starting from a certain time prior to the license expiration, as well as post-expiration.

## Scope

Self-hosted Teleport will be the target of this RFD. Teleport Cloud is out of scope because the Cloud license does not have proper expiration yet. When Cloud implements proper license expiration, this feature will kick in automatically.
License warning should be displayed in both Web UI and CLI (tsh/tctl).

### CLI UX

License warnings should be displayed in CLI (tsh/tctl) 
TSH - on “tsh login”, in “tsh status”
When the user uses either the “tsh login” or ‘tsh status” commands the appropriate license warning will be displayed. (see examples below)
TCTL - “tctl status”
When the user uses the “tctl status” command the appropriate license warning will be displayed.
The warnings are to be displayed 90 days prior to the license expiring.

> After expiration all tctl commands will display the license expiry warning.

Author: I was under the impression it was mainly the tsh commands we were trying to avoid adding this extra latency to. Currently i have been working under the assumption that all tctl commands should be checking this. If the extra request adding delays is a big enough concern for tctl as well I could modify it to only check on tctl status similar to how we are handling tsh.

> After expiration all tctl commands will display the license expiry warning.

Author: I've reworded this section to be clearer

### Example outputs

The license warning portion of the outputs should be coloured red and yellow respectively for the expired and unexpired warnings.
This can be determined by the severity of the cluster alert, 5-9 for yellow and 10+ for red.

```
$ build/tsh status
Profile URL:        https://localhost:9876
Logged in as:       someUser
Cluster:            someCluster
Roles:              access, editor
Logins:             someUser
Valid until:        2022-08-12 21:57:19 +0100 IST [valid for 4h44m0s]
Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

> Your Teleport Enterprise Edition license will expire in 10 days. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.

Author: Ah sorry, minor formatting error. fixed it now

        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

```
{
  "alerts": [
    {
      "kind": "cluster_alert",
      "version": "v1",
      "metadata": {
        "name": "123e4567-e89b-12d3-a456-426614174000",
        "labels": {
          "teleport.internal/alert-on-login": "yes",
          "teleport.internal/alert-permit-all": "yes",
          "teleport.internal/license-warning-expired": "yes",
        },
        "expires": "2022-08-31T17:26:05.728149Z"
      },
      "spec": {
        "severity": 5,
        "message": "Your Teleport Enterprise Edition license has expired. Please reach out to [licenses@goteleport.com](mailto:licenses@goteleport.com) to obtain a new license. Inaction may lead to unplanned outage or degraded performance and support.",
        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

On startup and every 1 hour afterwards the auth server will check the license and generate a license warning if applicable, or clear license warning alerts if they are no longer needed.

> All requests to grab cluster alerts on login will be made with a timeout of 500ms.

Author: I've fixed this line to be clearer. All requests will have this timeout

        "created": "2022-08-30T17:26:05.728149Z"
      }
    }
  ]
}
```

On startup and every 1 hour afterwards the auth server will check the license and generate a license warning if applicable, or clear license warning alerts if they are no longer needed.

All requests to grab cluster alerts will be made with a timeout of 500ms.

These license warning alerts will need to need to know which instance the alert came from so this will need to be added to the cluster alert spec to allow this. The cluster alert spec may also need to be modified to add a bool for whether an alert is allowed to be dismissed.

```
message ClusterAlertSpec {
  // Severity represents how problematic/urgent the alert is.
  AlertSeverity Severity = 1 [(gogoproto.jsontag) = "severity"];
  // Message is the user-facing message associated with the alert.
  string Message = 2 [(gogoproto.jsontag) = "message"];
  // Created is the time at which the alert was generated.
  google.protobuf.Timestamp Created = 3 [
    (gogoproto.jsontag) = "created,omitempty",
    (gogoproto.stdtime) = true,
    (gogoproto.nullable) = false
  ];

>   // InstanceName is the name of the instance that generated the alert.

Author: At the moment there should be no case where the instnaceName is not the instance that generated it. In the future if that was needed I think adding an extra field for that separately would make more sense than having both cases covered by this field
